pkg/pillar/base/execwrapper.go

```diff
@@ -0,0 +1,209 @@
+		return nil, fmt.Errorf("execCommand(%v): error while starting command: %s", c.command.Args, err.Error())
+	}
+
+	// Use a channel to signal completion so we can use a select statement
+	done := make(chan error)
+	go func() { done <- c.command.Wait() }()
+	stillRunning := time.NewTicker(25 * time.Second)
+
+	timer := time.After(c.timeout * time.Second)
+	for {
+		select {
+		case <-timer:
+			// Timeout happened first, kill the process.
+			c.command.Process.Kill()
+			return nil, fmt.Errorf("execCommand(%v): command timed out", c.command.Args)
+		case err := <-done:
+			// Command completed before timeout.
+			return c.buffer.Bytes(), err
+		case <-stillRunning.C:
+		}
+		updateAgentTouchFile(c.log, c.agentName)
+	}
+}
+
+//Exec returns Command object
+func Exec(log *LogObject, command string, arg ...string) *Command {
+	return &Command{
+		command:   exec.Command(command, arg...),
+		log:       log,
+		agentName: getAgentName(),
+	}
+}
+
+//updateAgentTouchFile updates agent's touch file under /var/run/
+func updateAgentTouchFile(log *LogObject, agentName string) {
+	if agentName == "" {
+		if log != nil {
+			log.Warnf("updateAgentTouchFile: agentName is empty")
+		}
+		return
+	}
+	filename := fmt.Sprintf("/var/run/%s.touch", agentName)
+	_, err := os.Stat(filename)
+	if err != nil {
+		file, err := os.Create(filename)
+		if err != nil {
+			if log != nil {
+				log.Infof("updateAgentTouchFile: %s\n", err)
+			}
+			return
+		}
+		file.Close()
+	}
+	_, err = os.Stat(filename)
+	if err != nil {
+		if log != nil {
+			log.Errorf("updateAgentTouchFile: %s\n", err)
+		}
+		return
+	}
+	now := time.Now()
+	err = os.Chtimes(filename, now, now)
+	if err != nil {
+		if log != nil {
+			log.Errorf("updateAgentTouchFile: %s\n", err)
+		}
+		return
+	}
+}
+
+//getAgentName parses stacktrace and returns involved agent under /pillar/cmd.
+func getAgentName() string {
+	stackTrace := strings.Split(string(debug.Stack()), "\n")
+	var methodName, file, agent, resultAgentName string
+	for i, trace := range stackTrace {
+		if strings.HasPrefix(trace, "github.com/lf-edge/eve/pkg/pillar/cmd/") {
+			trace = strings.TrimPrefix(trace, "github.com/lf-edge/eve/pkg/pillar/cmd/")
+			start := strings.Index(trace, ".")
+			end := strings.Index(trace, "(")
+			if start > -1 && end > -1 {
+				methodName = strings.TrimSpace(trace[start+1 : end])
+				agent = strings.TrimSpace(trace[:start])
+			}
+			filePath := strings.TrimSpace(stackTrace[i+1])
+			start = strings.LastIndex(filePath, "/")
+			end = strings.Index(filePath, ":")
+			if start > -1 && end > -1 {
+				file = strings.TrimSpace(filePath[start+1 : end])
+			}
+			break
+		}
+	}
+	switch agent {
+	case "zedagent":
+		switch file {
+		case "handlecertconfig.go":
+			if methodName == "getCertsFromController" || methodName == "parseControllerCerts" {
+				resultAgentName = "zedagentccerts"
+			} else {
+				resultAgentName = "zedagentattest"
+			}
+		case "handleinfo.go":
+			resultAgentName = "zedagentdevinfo"
```

Comment:
Will this need to change?

Reply:
Is it better to only use the methodName for zedagent i.e., not use the file?

Reply:
>Will this need to change?

Done.

>Is it better to only use the methodName for zedagent i.e., not use the file?

That might lead to having multiple switch cases to figure out an agentName. By having filename, I thought in future even if we add a method to a file which does an exec call, then we don't need to touch this code.

Reply:
The problem is that you have more cases when things will silently break. So we need to limit any refactoring to zedagent until we have moved out the exec calls.